crates/telio-pmtu/src/linux.rs

```diff
@@ -0,0 +1,377 @@
+
+        let mut msgbuf = vec![0xFF; super::PMTU_RANGE.end as usize];
+
+        if self.host.is_ipv6() {
+            let mut pkg = IcmpV6::new(&mut msgbuf)?;
+            self.probe_pmtu_internal(&mut pkg, pmtu).await
+        } else {
+            let mut pkg = IcmpV4::new(&mut msgbuf)?;
+            self.probe_pmtu_internal(&mut pkg, pmtu).await
+        }
+    }
+
+    async fn probe_pmtu_internal(&self, pkg: &mut impl IcmpPkg, mut pmtu: u32) -> io::Result<u32> {
+        pkg.set_ident(super::ICMP_ECHO_REQ_IDENT);
+
+        let mut pmtu_range = super::PMTU_RANGE;
+
+        loop {
+            let check = self.is_pmtu_ok(pkg, pmtu).await?;
+            telio_log_debug!("PMTU {pmtu} result: {check:?}");
+
+            pmtu = match check {
+                PmtuResult::Ok => {
+                    pmtu_range.start = pmtu;
+                    (pmtu_range.start + pmtu_range.end) / 2
+                }
+                PmtuResult::Dropped => {
+                    pmtu_range.end = pmtu;
+                    (pmtu_range.start + pmtu_range.end) / 2
+                }
+                PmtuResult::FragNeeded => {
+                    let mtu = read_error_pmtu(&self.sock, self.host.is_ipv6())?;
+                    telio_log_debug!("Kernel estimate of pmtu: {mtu}");
+
+                    if mtu != pmtu {
+                        // Sometimes kernel return error for packet of size equal to MTU
+                        pmtu_range.end = pmtu;
+                    }
+
+                    if mtu < pmtu_range.start {
+                        // Don't trust low values of nexthop MTU
+                        (pmtu_range.start + pmtu_range.end) / 2
+                    } else if pmtu == mtu {
+                        mtu + 1
+                    } else if pmtu == mtu + 1 {
+                        (pmtu_range.start + pmtu_range.end) / 2
+                    } else {
+                        mtu
+                    }
+                }
+            };
+
+            if pmtu_range.start + 1 >= pmtu_range.end {
+                break;
+            }
+        }
+
+        let found = pmtu_range.start;
+
+        if found == super::PMTU_RANGE.start {
+            Err(io::ErrorKind::NotConnected.into())
+        } else {
+            Ok(found)
+        }
+    }
+
+    async fn is_pmtu_ok(&self, pkg: &mut impl IcmpPkg, pmtu: u32) -> io::Result<PmtuResult> {
+        let mut recvbuf = [0u8; 1024];
+
+        for count in 0..RETRIES {
+            pkg.advance_seq();
+            let seq = pkg.seq();
+
+            telio_log_debug!(
+                "Sending ICMP Echo Request, dst: {} seq: {seq}, size: {pmtu}",
+                self.host
+            );
+
+            if let Err(err) = self.sock.send(pkg.bytes(pmtu)).await {
+                // The kernel might retun this error when the packet size == kernel's estimate of PMTU.
+                // But the packet is sent anyway
+                if err.raw_os_error() != Some(libc::EMSGSIZE) {
```

Comment:
Suggestion: instead of the comment, we can leave the URL to https://man7.org/linux/man-pages/man7/udp.7.html
and copy an excerpt:
```
       By default, Linux UDP does path MTU (Maximum Transmission Unit)
       discovery.  This means the kernel will keep track of the MTU to a
       specific target IP address and return EMSGSIZE when a UDP packet
       write exceeds it.  When this happens, the application should
       decrease the packet size.
```

Reply:
This passage is not relevant here and we operate on ICMP protocol

Reply:
My bad. The correct excerpt seems to be this one:
```
    If the message is too long to pass atomically through the
       underlying protocol, the error EMSGSIZE is returned, and the
       message is not transmitted.
```

from the https://man7.org/linux/man-pages/man2/send.2.html

My point being that comment tells me that despite the error it was ommited but there doesn't seem to be a manual entry guarding that. Linking an appropriate documentation would help a lot in this case.

Of course if what's happening is not what's described in the manual then we have another problem.

Reply:
That makes things a bit more complicated but I guess that's just how it is. Thanks for explaining